### PR TITLE
Add column field ID control in parquet writer

### PR DESCRIPTION
**Pull request by PointKernel**
Closes https://github.com/rapidsai/cudf/issues/10375
Closes https://github.com/rapidsai/cudf/issues/10376

This PR enables column `field_id` control in the parquet writer. When writing a parquet file, users can specify a column's `field_id` via `column_in_metadata.set_parquet_field_id()`. JNI bindings and uni tests are added as well.

### CODE REVIEW

File: cpp/src/io/parquet/parquet.hpp
Comment (jlowe) [inline cpp/src/io/parquet/parquet.hpp:21]: This include does not seem to be used?

File: cpp/src/io/parquet/writer_impl.cu
Comment (devavret) [inline cpp/src/io/parquet/writer_impl.cu:537]: remove the default capture if not needed

File: cpp/src/io/parquet/parquet.hpp
Comment (devavret) [reply cpp/src/io/parquet/parquet.hpp:165]: You can store an optional member in this struct. Then the equality check will just work

**Review by devavret — COMMENTED** [review]
Looks fine. Apart from @jlowe's concerns, this should be good to go

File: cpp/src/io/parquet/parquet.hpp
Comment (PointKernel) [reply cpp/src/io/parquet/parquet.hpp:165]: I was using an optional member in the first place and the leftover of including thrust optional at the top of this file can prove this :smile: . The problem is during [reading](
https://github.com/PointKernel/cudf/blob/parquet-field-id-writing/cpp/src/io/parquet/compact_protocol_reader.cpp#L159), where we need to read the value of the optional and the optional can have no value:
```cpp
bool CompactProtocolReader::read(SchemaElement* s)
{
  auto op = std::make_tuple(ParquetFieldEnum<Type>(1, s->type),
                        ParquetFieldInt32(2, s->type_length),
                        ParquetFieldEnum<FieldRepetitionType>(3, s->repetition_type),
                        ParquetFieldString(4, s->name),
                        ParquetFieldInt32(5, s->num_children),
                        ParquetFieldEnum<ConvertedType>(6, s->converted_type),
                        ParquetFieldInt32(7, s->decimal_scale),
                        ParquetFieldInt32(8, s->decimal_precision),
                        ParquetFieldInt32(9, s->field_id.value()), // read a nullable value
                        ParquetFieldStruct(10, s->logical_type));
  return function_builder(this, op);
}
```
We don't know whether `field_id` has been set or not beforehand so when `field_id` is not set, execution will throw:
```
C++ exception with description "Optional has no value" thrown in the test body.
```
I haven't figured out how to handle a `thrust::optional` with the current compact protocol reader. I guess a new optional field reader is required? By all means, I'm removing the changes to the compact protocol reader so it won't read `field_id` for now.

**Review by jlowe — DISMISSED** [review]

**Review by jlowe — APPROVED** [review]

File: cpp/tests/io/parquet_test.cpp
Comment (vuule) [inline cpp/tests/io/parquet_test.cpp:3230]:
```suggestion
  EXPECT_TRUE(got_metadata->column_metadata[0].is_parquet_field_id_set());
```

File: cpp/tests/io/parquet_test.cpp
Comment (vuule) [inline cpp/tests/io/parquet_test.cpp:3212]: Do we need rows for this test, could we use empty columns?

File: cpp/src/io/parquet/writer_impl.cu
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:539]: why is the list check required?

File: cpp/src/io/parquet/parquet.hpp
Comment (vuule) [reply cpp/src/io/parquet/parquet.hpp]: I would be really nice to read the field_id values in the reader. I think we can add `ParquetFieldOptionalInt32` for this purpose, or even turn `ParquetFieldInt32` into a template so it can also hold a reference to an optional. I think implementation is the same.
This is the exact reason why I had to refactor the ORC equivalent and use templates/overloads.

Comment (vuule) [reply cpp/src/io/parquet/parquet.hpp]: I was thinking to only use it for the field ID for now. It bugs me that we can't round trip an ID. I think the limited tests demonstrate that the feature is incomplete. 
Not blocking the PR on this, but would prefer to have minimal reader changes included.

Comment (PointKernel) [reply cpp/src/io/parquet/parquet.hpp]: Got it. Will do.

File: cpp/src/io/parquet/writer_impl.cu
Comment (devavret) [reply cpp/src/io/parquet/writer_impl.cu:539]: presumably because "stub" elements can't have a field id

Comment (PointKernel) [reply cpp/src/io/parquet/writer_impl.cu:539]: Similar to how we set schema names, my idea was to not set field ID for "intermediate" schemas for lists. Removing it now since it's not necessarily required.

Comment (PointKernel) [reply cpp/src/io/parquet/writer_impl.cu:539]: Based on offline discussion with @devavret, we agreed to keep the list check since "stub" elements are not supposed to have field ID.

Comment (devavret) [reply cpp/src/io/parquet/writer_impl.cu:539]: Let's confirm it as well. @jlowe , would spark have a `field_id` corresponding to a stub element? Or is it mostly used for the parent-most level?

Comment (PointKernel) [reply cpp/src/io/parquet/writer_impl.cu:539]: Based on the PR description in https://issues.apache.org/jira/browse/SPARK-38094:
```
After this PR, when the requested schema has field IDs, Parquet readers will first use the field ID to determine which Parquet columns to read, before falling back to using column names as before.
```
I assume it's mainly for the outter-most level.

File: cpp/tests/io/parquet_test.cpp
Comment (vuule) [inline cpp/tests/io/parquet_test.cpp:3207]: we should now be able to test that IDs can survive a round trip through a Parquet file.

File: cpp/include/cudf/io/types.hpp
Comment (vuule) [inline cpp/include/cudf/io/types.hpp:398]: as discussed offline, can't use optional here because of Cython

**Review by vuule — CHANGES_REQUESTED** [review]
Looks good. Just a few things to reach 🔥 status.

File: cpp/tests/io/parquet_test.cpp
Comment (PointKernel) [reply cpp/tests/io/parquet_test.cpp:3207]: Currently bypass the round trip test since the reader-side metadata is not exposed.

**Review by vuule — APPROVED** [review]
LGTM
Thank you for iterating on the ideas that came up in the review.

**Comment by PointKernel** [comment]
@jlowe Just want to make sure, would this PR resolve your issue?

File: cpp/src/io/parquet/writer_impl.cu
Comment (jlowe) [reply cpp/src/io/parquet/writer_impl.cu:539]: > would spark have a field_id corresponding to a stub element?

It is expecting to find a specified field ID on a child column of a StructType (STRUCT) column.  Theoretically there could be a STRUCT of LIST (array) column and the user could specify it on the array column.  See https://github.com/apache/spark/pull/35385/files#diff-487304e31da0dcde467c1f8561f42edcb3a811a755d8bc0424e4f3ad084099c3R156-R175

However I don't know whether the underlying Parquet message types allow a field ID on the array column vs. the child column of the array.

**Comment by jlowe** [comment]
> @jlowe Just want to make sure, would this PR resolve your issue?

From an API perspective it should address what we need.  Seems like there's still a question around the list column exclusion in the implementation and whether that check is needed in practice.

**Comment by res-life** [comment]
I tested, it works, you can merge it or I'll add the JNI code into this PR.
```
Properties:

org.apache.spark.version: 3.3.0
org.apache.spark.sql.parquet.row.metadata: 
{
	"type": "struct",
	"fields": [{
		"name": "c1",
		"type": "integer",
		"nullable": true,
		"metadata": {
			"parquet.field.id": 1
		}
	}, {
		"name": "c2",
		"type": {
			"type": "struct",
			"fields": [{
				"name": "c3",
				"type": "integer",
				"nullable": true,
				"metadata": {
					"parquet.field.id": 3
				}
			}]
		},
		"nullable": true,
		"metadata": {
			"parquet.field.id": 2
		}
	}]
}

Schema:
message schema {
  optional int32 c1 = 1;
  optional group c2 = 2 {
    optional int32 c3 = 3;
  }
}


```

**Comment by PointKernel** [comment]
@res-life Thanks! I just gave you the write access to my repo. Can you please add the JNI bindings and tests to this PR?

**Comment by res-life** [comment]
@jlowe help review JNI part

File: cpp/tests/io/parquet_test.cpp
Comment (devavret) [inline cpp/tests/io/parquet_test.cpp:3225]: I suppose this test isn't needed anymore?

**Review by devavret — COMMENTED** [review]
Almost there!

File: cpp/tests/io/parquet_test.cpp
Comment (PointKernel) [reply cpp/tests/io/parquet_test.cpp]: Removed

**Review by devavret — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:37]: How does one specify they do _not_ want a field ID written for this column?  It looks like we're always going to write everything with a field ID of zero unless something is specified.  We should only specify a field ID to Parquet when it is specified as a column writer option explicitly.  I'm not sure if there's a safe sentinel value to use here, so we may need another boolean to indicate whether this field has been set and is valid to use.

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:7904]: There should be a test that we are doing the correct thing when _not_ specfying a field ID to use.

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java
Comment (res-life) [reply java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:37]: Negative field ID is valid, so should add another boolean as you mentioned.

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (res-life) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:7904]: Added a not specified `c5` column in the existing test case.

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java
Comment (res-life) [reply java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:38]: done.

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:7910]: Would be good to add a negative test for a nested column, as it's a slightly different code path that appears to be problematic in the current version of this PR.

File: java/src/main/native/src/TableJni.cpp
Comment (res-life) [reply java/src/main/native/src/TableJni.cpp:693]: Yes, it's a bug. Updated.

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (res-life) [reply java/src/test/java/ai/rapids/cudf/TableTest.java:7910]: Added a test case.

Spark allows two columns have identical `parquet.field.id`, do did not test this kind of case.

```
from pyspark.sql.types import *
schema = StructType([
    StructField("c1", IntegerType(), metadata={'parquet.field.id' : 1}),
    StructField("c2", IntegerType(), metadata={'parquet.field.id' : 1})
])
data = [(1,2,),(11,22,),(33,33,),]
spark.createDataFrame(data, schema).coalesce(1).write.mode("overwrite").parquet("/tmp/tmp-parquet")
```

parquet-tool meta /tmp/tmp-parquet/xxx.parquet

```
......
Schema:
message spark_schema {
  optional int32 c1 = 1;
  optional int32 c2 = 1;
}
```

**Review by jlowe — APPROVED** [review]

**Comment by PointKernel** [comment]
@gpucibot merge

**Pull request merged**
